Fix unix socket subpath mount

gnufied opened the pull request:
Fixes https://github.com/kubernetes/kubernetes/issues/61377

```release-note
Fix mounting of UNIX sockets(and other special files) in subpaths
```

gnufied commented:
/sig storage

liggitt commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1131,18 +1172,38 @@ func doSafeOpen(pathname string, base string) (int, error) {
+					iNodeNumber:   deviceStat.Ino,
+					specialDevice: true,
+				}
+				return -1, deviceInfo, nil
```
returning here is incorrect if this is not the final segment (e.g. if the path was /foo/bar/baz and /foo/bar is a special file, we cannot descend to check the `baz` child and should fail)

gnufied replied:
good point. should be easy to fix.

liggitt commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1131,18 +1172,38 @@ func doSafeOpen(pathname string, base string) (int, error) {
 		childFD, err = syscall.Openat(parentFD, seg, nofollowFlags, 0)
 		if err != nil {
-			return -1, fmt.Errorf("cannot open %s: %s", currentPath, err)
+			if err == syscall.ENXIO {
```
structurally, prefer this unnested:
```
switch {
case err == syscall.ENXIO:
  ...
case err != nil:
  ...
}
```

@kubernetes/sig-storage-pr-reviews will need to weigh in on whether this handling is correct

liggitt commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1131,18 +1172,38 @@ func doSafeOpen(pathname string, base string) (int, error) {
+
+				fileFmt := deviceStat.Mode & syscall.S_IFMT
+
+				if fileFmt == syscall.S_IFREG || fileFmt == syscall.S_IFDIR {
```
would it be better to check against a whitelist of allowed/expected types? for example, this seems to allow `syscall.S_IFLNK`

gnufied replied:
We can but in odd case where it does happen, what will happen is - inode and device number capture here will belong to symlink itself, whereas bindTarget will resolve the symlink target and have different inode and device number

msau42 replied:
Yeah, prefer whitelisting allowed types and erroring on the rest.  That way we don't accidentally allow something that shouldn't be.

msau42 commented on pkg/util/mount/mount_linux.go:
```diff
@@ -790,11 +796,25 @@ func doBindSubPath(mounter Interface, subpath Subpath, kubeletPid int) (hostPath
+	}
+
+	if deviceInfo.specialDevice {
+		if err = mounter.Mount(evalSubPath, bindPathTarget, "" /*fstype*/, []string{"bind"}); err != nil {
```
Also add in the logging

msau42 reviewed (commented):
Tests?

gnufied commented:
@msau42 @liggitt fixed comments. Also added review comments.

BTW - I changed the point at which `Openat` is called. I noticed call to `Openat` can hang for FIFOs, so it is best to call `fstatat` before calling `Openat`.

liggitt commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1125,24 +1169,44 @@ func doSafeOpen(pathname string, base string) (int, error) {
+				}
+				return -1, deviceInfo, nil
+			}
+			return -1, defaultDeviceInfo, fmt.Errorf("cannot open %s: %s", currentPath, err)
```
better error here about a subpath being a special device

msau42 commented on pkg/util/mount/mount_linux_test.go:
```diff
@@ -1179,6 +1180,25 @@ func TestBindSubPath(t *testing.T) {
 			expectError: false,
 		},
+		{
+			name: "subpath-mounting-unix-socket",
```
Also a testcase where socket is not the last segment in the path.

msau42 replied:
Also testcase for fifo?

gnufied replied:
added tests. also added test for fifo

msau42 commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1125,24 +1169,44 @@ func doSafeOpen(pathname string, base string) (int, error) {
+			return -1, defaultDeviceInfo, fmt.Errorf("path %s is outside of allowed base %s", currentPath, base)
+		}
+		var deviceStat unix.Stat_t
+		deviceStatErr := unix.Fstatat(parentFD, seg, &deviceStat, unix.AT_SYMLINK_NOFOLLOW)
```
I think SafeMakeDir also needs checks for these.  findExistingPrefix also does Openat in a loop

gnufied replied:
So I think if we are going to use `Fstatat` in `findExistingPrefix` then there is no need to use `Openat` there because all we are interesting in knowing is - first directory that does not exist. We don't need `Openat` for that.

tallclair commented on Godeps/Godeps.json:
```diff
@@ -2350,82 +2348,82 @@
 		{
 			"ImportPath": "github.com/opencontainers/runc/libcontainer/utils",
-			"Comment": "v1.0.0-rc4-221-g595bea02",
+			"Comment": "v1.0.0-rc4-221-g595bea0",
```
This is updating a lot more than `sys/unix`. Are all these other changes intentional? These hashes look suspicious.

liggitt replied:
Added another significant digit. Not sure if that's a git thing or a godep thing

gnufied replied:
Yeah - the SHAs are same, it is just updating `Comment` field too, not the actual `Rev` field. I am not sure if it was because of git or godep either. :(

gnufied replied:
I can probably manually drop these hunks and just keep the bits I need. Although - I am not sure if full restore, edit and save will again cause these `Comment` fields to be update for someone.

gnufied commented:
/retest

gnufied committed:
Use O_PATH to avoid errors on Openat

Also add tests for sockets and fifos
Replace use of syscall for flags with unix

gnufied commented:
@liggitt @msau42 much simpler version of patch that uses `O_PATH` and still works. PTAL.

liggitt commented on pkg/util/mount/mount_linux_test.go:
```diff
@@ -1435,6 +1475,32 @@ func TestSafeOpen(t *testing.T) {
+		{
+			"mounting-unix-socket",
+			func(base string) error {
+				testSocketFile := filepath.Join(base, "mount_test.sock")
```
do you need to clean this up on exit?

gnufied replied:
Same as below, since we are removing the base directory via https://github.com/kubernetes/kubernetes/blob/master/pkg/util/mount/mount_linux_test.go#L658 , it should remove any files created in the tests.

liggitt commented on pkg/util/mount/mount_linux_test.go:
```diff
@@ -1435,6 +1475,32 @@ func TestSafeOpen(t *testing.T) {
+		{
+			"mounting-unix-socket-in-middle",
+			func(base string) error {
+				testSocketFile := filepath.Join(base, "mount_test.sock")
```
same here?

liggitt commented on pkg/util/mount/mount_linux_test.go:
```diff
@@ -1551,6 +1617,17 @@ func TestFindExistingPrefix(t *testing.T) {
+		{
+			"with-fifo-in-middle",
+			func(base string) error {
+				testFifo := filepath.Join(base, "mount_test.fifo")
```
cleanup?

gnufied replied:
the test removes entire base dir on exit.

gnufied commented:
/test pull-kubernetes-integration

gnufied commented on pkg/util/mount/mount_linux.go:
```diff
@@ -1074,7 +1076,9 @@ func findExistingPrefix(base, pathname string) (string, []string, error) {
-		childFD, err := syscall.Openat(fd, dir, syscall.O_RDONLY, 0)
+		// Using O_PATH here will prevent hangs in case user replaces directory with
+		// fifo
+		childFD, err := syscall.Openat(fd, dir, unix.O_PATH, 0)
```
yeah you can't use `read` on a file descriptor opened with `O_PATH` and hence `O_RDONLY` will be ignored.

gnufied commented on pkg/util/mount/mount_linux.go:
```diff
@@ -55,6 +55,8 @@ const (
 	// syscall.Openat flags used to traverse directories not following symlinks
 	nofollowFlags = syscall.O_RDONLY | syscall.O_NOFOLLOW
+	// flags for getting file descriptor without following the symlink
+	openFDFlags = syscall.O_NOFOLLOW | unix.O_PATH
```
Both the flags are described in `man 2 open` . It is unfortunate that golang goes not already include `O_PATH` flag in `syscall` package and one must use `golang.org/x/sys/unix` .

liggitt replied:
Can you open a golang issue for that?

msau42 replied:
I think it's an issue on us.  syscall package is deprecated, and golang suggests everyone use the unix package instead

liggitt replied:
Ah. Great.

gnufied replied:
replaced usage of `syscall` with `unix` for these flags.

msau42 replied:
I meant there's a bigger issue on us.  We should replace usage of the entire syscall package with unix package (but probably not in this pr...)

msau42 commented:
/lgtm

msau42 commented on pkg/util/mount/mount_linux.go:
```diff
@@ -54,7 +54,9 @@ const (
 	containerSubPathDirectoryName = "volume-subpaths"
 	// syscall.Openat flags used to traverse directories not following symlinks
-	nofollowFlags = syscall.O_RDONLY | syscall.O_NOFOLLOW
+	nofollowFlags = unix.O_RDONLY | unix.O_NOFOLLOW
```
Do we still use this anywhere?

gnufied replied:
Yes, in few places with outer `Open` calls when we are opening base directory.

msau42 commented:
/lgtm

gnufied commented:
assigning @saad-ali for approval

/assign @saad-ali

saad-ali commented:
/approve

saad-ali commented:
Cherry pick to 1.10 (after 1.10.0 release), and other relevant branches (1.9, etc.).

gnufied commented:
@saad-ali accidently removed "approve" label?
